proposals/utils/proposal_utils.py

```diff
@@ -373,3 +377,69 @@ def notify_local_staff(proposal):
 
     # Reset the current language
     activate(current_language)
+
+
+@deconstructible
+class FilenameFactory:
+    '''A callable class which can be passed to upload_to() in FileFields
+    and can be deconstructed for migrations'''
+    
+    def __init__(self, document_type):
+        self.document_type = document_type
+    
+    def __call__(self, instance, original_fn):
+        '''Returns a custom filename preserving the original extension,
+        something like "FETC-2020-002-01-Villeneuve-T2-Informed-Consent.pdf"'''
+        
+        # Importing here to prevent circular import
+        from proposals.models import Proposal
+        
+        if isinstance(instance, Proposal):
+            proposal = instance
+            trajectory = None
+        else:
+            # In case of Documents or Study objects
+            proposal = instance.proposal
+            try:
+                trajectory = 'T' + str(instance.study.order)
+            except AttributeError:
+                trajectory = None
```

Comment:
Just realised, if a researcher adds more than 1 extra document of the same type, this won't give both of them an unique name.

Reply:
As for a fix, maybe do a QS of extra document objects and see which index this object is in that QS?

Reply:
Good tip. I probably should have caught this, too.

Reply:
That's why I wanted to do these reviews. I make these kinds of oversights all the time...